### ISSUE TITLE
`log_softmax` could be `2**103` to `2**1022` times more accurate on small outputs

### ISSUE THREAD

**Issue by JasonGross** [pytorch/pytorch#113708]
### 🚀 The feature, motivation and pitch

As I understand it, PyTorch implements `log_softmax(x)` as `x - x.max() - (x - x.max()).exp().sum().log()`. However, when the largest value is much larger than the rest of the values (about 16 larger for float32, about 36 larger for float64), `log_softmax` returns `0` at the maximum value, when it could give a much more precise answer.

Numerically, we have
```python
>>> eps = torch.tensor(torch.finfo(torch.float32).eps, dtype=torch.float32)
>>> -torch.tensor([1-(2*eps).log(), 0]).log_softmax(dim=0)
tensor([1.1921e-07, 1.6249e+01])
>>> -torch.tensor([1-eps.log(), 0]).log_softmax(dim=0)
tensor([-0.0000, 16.9424])
```

(For float16, this is 7.2e-4, for float64 2.2e-16.)

The value of `(x - x.max()).exp().sum()` is going to be at least 1, because the largest value of `x - x.max()` is 0.  1.19e-7 is the unit of least precision for float32 around 1, but float32 can represent values as small as about 1.18e-38 normally (`torch.finfo(torch.float32).smallest_normal`, `2**-126 = 2**-(2**7 - 2)`) or about 1.18e-45 (`2**-126 * 2**-23`) subnormally.  So there's a lot of room for more precision using `log1p(x) = log(1 + x)` instead.  Consider:
```python
eps = torch.tensor(torch.finfo(torch.float32).smallest_normal, dtype=torch.float32)
x = torch.tensor([1-eps.log(), 0])
maxi = x.argmax()
xoffset = x - x[maxi]
xoffsetexp = xoffset.exp()
# xoffsetexp[maxi] is current about 1
xoffsetexp[maxi] = 0
xoffsetexp_sum_m1 = xoffsetexp.sum()
print(-(xoffset - xoffsetexp_sum_m1.log1p())) # tensor([4.3244e-39, 8.8337e+01])
print((-(xoffset - xoffsetexp_sum_m1.log1p())[0]).log2()) # tensor(-127.4427)
```

For float64 and using `smallest_subnormal` instead of `smallest_normal`, the difference is even more pronounced:
```python
eps = torch.tensor(numpy.finfo(numpy.float64).smallest_subnormal, dtype=torch.float64)
x = torch.tensor([1-(8*eps).log(), 0])
maxi = x.argmax()
xoffset = x - x[maxi]
xoffsetexp = xoffset.exp()
# xoffsetexp[maxi] is current about 1
xoffsetexp[maxi] = 0
xoffsetexp_sum_m1 = xoffsetexp.sum()
print(-(xoffset - xoffsetexp_sum_m1.log1p())) # tensor([1.9763e-323,  7.4336e+02], dtype=torch.float64)
print((-(xoffset - xoffsetexp_sum_m1.log1p())[0]).log2()) # tensor(-1072., dtype=torch.float64)
```

This came up when a transformer I was training with cross-entropy loss on a classification task had loss dominated by `torch.finfo(torch.float32).eps`.

### Additional context

I originally posted this as a [StackOverflow question](https://stackoverflow.com/q/77477327/377022).

**Comment by janeyx99** [comment]
cc @drisspg

**Comment by abcde2342** [comment]
I think we can have a patch based on Taylor series:

```
def log1p_taylor_scalar(s, terms=8):
    total = 0.0
    power = 1.0
    for n in range(1, terms + 1):
        power *= s
        total += (-1.0) ** (n + 1) * power / n
    return total

def log_softmax_taylor(x, terms=8):
    x = np.asarray(x)
    i = int(np.argmax(x))
    off = x - x[i]
    ex = np.exp(off)
    ex[i] = 0
    s = float(np.sum(ex))
    return off - log1p_taylor_scalar(s, terms=terms)
```
